mlflow/projects/utils.py

```diff
@@ -80,3 +100,186 @@ def _is_valid_branch_name(work_dir, version):
+    response = requests.get(uri)
+    try:
+        response.raise_for_status()
+    except requests.HTTPError as error:
+        raise ExecutionException("Unable to retrieve ZIP file. Reason: %s" % str(error))
+    return BytesIO(response.content)
+
+
+def get_or_create_run(run_id, uri, experiment_id, work_dir, entry_point):
+    if run_id:
+        return tracking.MlflowClient().get_run(run_id)
+    else:
+        return _create_run(uri, experiment_id, work_dir, entry_point)
+
+
+def _create_run(uri, experiment_id, work_dir, entry_point):
+    """
+    Create a ``Run`` against the current MLflow tracking server, logging metadata (e.g. the URI,
+    entry point, and parameters of the project) about the run. Return an ``ActiveRun`` that can be
+    used to report additional data about the run (metrics/params) to the tracking server.
+    """
+    if _is_local_uri(uri):
+        source_name = tracking._tracking_service.utils._get_git_url_if_present(_expand_uri(uri))
+    else:
+        source_name = _expand_uri(uri)
+    source_version = _get_git_commit(work_dir)
+    existing_run = fluent.active_run()
+    if existing_run:
+        parent_run_id = existing_run.info.run_id
+    else:
+        parent_run_id = None
+
+    tags = {
+        MLFLOW_USER: _get_user(),
+        MLFLOW_SOURCE_NAME: source_name,
+        MLFLOW_SOURCE_TYPE: SourceType.to_string(SourceType.PROJECT),
+        MLFLOW_PROJECT_ENTRY_POINT: entry_point
+    }
+    if source_version is not None:
+        tags[MLFLOW_GIT_COMMIT] = source_version
+    if parent_run_id is not None:
+        tags[MLFLOW_PARENT_RUN_ID] = parent_run_id
+
+    active_run = tracking.MlflowClient().create_run(experiment_id=experiment_id, tags=tags)
+    return active_run
+
+
+def log_project_params_and_tags(run, project, work_dir, entry_point, parameters, version):
+    # Consolidate parameters for logging.
+    # `storage_dir` is `None` since we want to log actual path not downloaded local path
+    entry_point_obj = project.get_entry_point(entry_point)
+    final_params, extra_params = entry_point_obj.compute_parameters(parameters, storage_dir=None)
+    for key, value in (list(final_params.items()) + list(extra_params.items())):
+        tracking.MlflowClient().log_param(run.info.run_id, key, value)
+
+    repo_url = _get_git_repo_url(work_dir)
+    if repo_url is not None:
+        for tag in [MLFLOW_GIT_REPO_URL, LEGACY_MLFLOW_GIT_REPO_URL]:
+            tracking.MlflowClient().set_tag(run.info.run_id, tag, repo_url)
+
+    # Add branch name tag if a branch is specified through -version
+    if _is_valid_branch_name(work_dir, version):
+        for tag in [MLFLOW_GIT_BRANCH, LEGACY_MLFLOW_GIT_BRANCH_NAME]:
+            tracking.MlflowClient().set_tag(run.info.run_id, tag, version)
+
+
+FetchedProject = collections.namedtuple("FetchedProject", ["project", "work_dir", "run"])
+
+
+def prepare_project_and_run(project_uri, experiment_id, run_id, entry_point, parameters, version):
```

Comment:
Would be good to add an end-to-end test for this method btw, in tests/projects/utils.py. I also wonder if it'd be easier for plugin implementors if we just exposed the component utility functions that comprise this method (`fetch_and_validate_project`, `get_or_create_run`, `log_project_params_and_tags`) and showed an example of how to call the three functions as we do here in order to fetch the project, create a run, and set appropriate tags - it may be confusing for plugin implementors to reason about what exactly happens in this method & its NamedTuple return value. What do you think?

Reply:
I remove the global method and created a test to validate the calls of these 3 methods.